# 데이터 불러오기

In [ ]:
# 라이브러리 불러오기
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

In [ ]:
# 데이터 불러오기
url = "https://bakey-api.codeit.kr/api/files/resource?root=static&seqId=10713&version=1&directory=/cosmetic.csv&name=cosmetic.csv"
cosmetic_df = pd.read_csv(url)

In [ ]:
# 데이터 확인하기
cosmetic_df.head()

,event_time,event_type,product_id,category_id,category_code,brand,price,user_id,user_session
0,2019-12-01 00:00:00 UTC,remove_from_cart,5712790,1487580005268456287,NaN,f.o.x,6.27,576802932,51d85cb0-897f-48d2-918b-ad63965c12dc
1,2019-12-01 00:00:00 UTC,view,5764655,1487580005411062629,NaN,cnd,29.05,412120092,8adff31e-2051-4894-9758-224bfa8aec18
2,2019-12-01 00:00:02 UTC,cart,4958,1487580009471148064,NaN,runail,1.19,494077766,c99a50e8-2fac-4c4d-89ec-41c05f114554
3,2019-12-01 00:00:05 UTC,view,5848413,1487580007675986893,NaN,freedecor,0.79,348405118,722ffea5-73c0-4924-8e8f-371ff8031af4
4,2019-12-01 00:00:07 UTC,view,5824148,1487580005511725929,NaN,NaN,5.56,576005683,28172809-7e4a-45ce-bab0-5efa90117cd5


In [ ]:
cosmetic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3533286 entries, 0 to 3533285
Data columns (total 9 columns):
 #   Column         Dtype  
---  ------         -----  
 0   event_time     object 
 1   event_type     object 
 2   product_id     int64  
 3   category_id    int64  
 4   category_code  object 
 5   brand          object 
 6   price          float64
 7   user_id        int64  
 8   user_session   object 
dtypes: float64(1), int64(3), object(5)
memory usage: 242.6+ MB


In [ ]:
# object -> datetime
cosmetic_df['event_time'] = pd.to_datetime(cosmetic_df['event_time'])

In [ ]:
cosmetic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3533286 entries, 0 to 3533285
Data columns (total 9 columns):
 #   Column         Dtype              
---  ------         -----              
 0   event_time     datetime64[ns, UTC]
 1   event_type     object             
 2   product_id     int64              
 3   category_id    int64              
 4   category_code  object             
 5   brand          object             
 6   price          float64            
 7   user_id        int64              
 8   user_session   object             
dtypes: datetime64[ns, UTC](1), float64(1), int64(3), object(4)
memory usage: 242.6+ MB


In [ ]:
cosmetic_df['event_time'].min(), cosmetic_df['event_time'].max()

(Timestamp('2019-12-01 00:00:00+0000', tz='UTC'),
 Timestamp('2019-12-31 23:59:57+0000', tz='UTC'))

In [ ]:
# event_type 범주의 종류
cosmetic_df['event_type'].unique()

array(['remove_from_cart', 'view', 'cart', 'purchase'], dtype=object)

In [ ]:
# event_type 확인하기 : view -> cart -> (remove_from_cart) -> purchase
cosmetic_df['event_type'].value_counts()

,count
event_type,
view,1728331
cart,927124
remove_from_cart,664655
purchase,213176


# 데이터 전처리

In [ ]:
cosmetic_df.shape

(3533286, 9)

## 결측값 확인

In [ ]:
cosmetic_df.isna().sum().sum()

np.int64(4985889)

In [ ]:
cosmetic_df.isna().sum()

,0
event_time,0
event_type,0
product_id,0
category_id,0
category_code,3474821
brand,1510289
price,0
user_id,0
user_session,779


In [ ]:
cosmetic_df.isna().mean()

,0
event_time,0.000000
event_type,0.000000
product_id,0.000000
category_id,0.000000
category_code,0.983453
brand,0.427446
price,0.000000
user_id,0.000000
user_session,0.000220


- category_code, brand, user_session 컬럼에 결측치가 존재하는 것을 확인할 수 있다.
- category_code의 경우 전체 데이터의 98.3%가 결측으로 분석에 활용하기가 어려워 category_code의 컬럼은 삭제하겠다.

In [ ]:
# category_code 컬럼 삭제
cosmetic_df = cosmetic_df.drop('category_code', axis=1)

In [ ]:
# brand 결측치 처리
cosmetic_df.query("product_id == 5916983")

,event_time,event_type,product_id,category_id,brand,price,user_id,user_session
3374835,2019-12-29 07:11:47+00:00,view,5916983,1783999073758478650,NaN,0.00,410144142,57484fc7-f6a4-45ba-b95a-9116ead6d96f
3397074,2019-12-29 13:00:50+00:00,cart,5916983,1783999073758478650,NaN,0.00,355374962,de91420a-9375-4477-b44f-c072e1604f2c
3397086,2019-12-29 13:00:55+00:00,remove_from_cart,5916983,1783999073758478650,NaN,0.00,355374962,de91420a-9375-4477-b44f-c072e1604f2c
3397087,2019-12-29 13:00:55+00:00,remove_from_cart,5916983,1783999073758478650,NaN,0.00,355374962,de91420a-9375-4477-b44f-c072e1604f2c
3397111,2019-12-29 13:01:02+00:00,view,5916983,1783999073758478650,NaN,0.00,355374962,de91420a-9375-4477-b44f-c072e1604f2c
3468885,2019-12-30 12:36:57+00:00,cart,5916983,1783999073758478650,beautific,27.52,592007896,84475418-083b-4a8e-ab7f-8d38994b1f12
3470599,2019-12-30 12:59:20+00:00,view,5916983,1783999073758478650,beautific,27.52,477270774,e5670c53-8674-4725-86e2-dc6bfd39c04c
3498570,2019-12-30 21:43:54+00:00,view,5916983,1783999073758478650,beautific,27.52,583884978,d6f873d5-7935-454c-9849-9fbf1b05fd53


- brand 컬럼의 결측치 처리 방법으로, 동일한 상품일 경우 브랜드 정보도 같으므로 product_id를 기준으로 결측치 대체를 고려할 수 있다.

In [ ]:
cosmetic_df['brand'] = cosmetic_df.groupby('product_id')['brand'].transform(lambda x: x.fillna(method='ffill').fillna(method='bfill'))

/tmp/ipykernel_1594/3342855910.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  cosmetic_df['brand'] = cosmetic_df.groupby('product_id')['brand'].transform(lambda x: x.fillna(method='ffill').fillna(method='bfill'))
/tmp/ipykernel_1594/3342855910.py:1: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  cosmetic_df['brand'] = cosmetic_df.groupby('product_id')['brand'].transform(lambda x: x.fillna(method='ffill').fillna(method='bfill'))


In [ ]:
cosmetic_df.query("product_id == 5916983")

,event_time,event_type,product_id,category_id,brand,price,user_id,user_session
3374835,2019-12-29 07:11:47+00:00,view,5916983,1783999073758478650,beautific,0.00,410144142,57484fc7-f6a4-45ba-b95a-9116ead6d96f
3397074,2019-12-29 13:00:50+00:00,cart,5916983,1783999073758478650,beautific,0.00,355374962,de91420a-9375-4477-b44f-c072e1604f2c
3397086,2019-12-29 13:00:55+00:00,remove_from_cart,5916983,1783999073758478650,beautific,0.00,355374962,de91420a-9375-4477-b44f-c072e1604f2c
3397087,2019-12-29 13:00:55+00:00,remove_from_cart,5916983,1783999073758478650,beautific,0.00,355374962,de91420a-9375-4477-b44f-c072e1604f2c
3397111,2019-12-29 13:01:02+00:00,view,5916983,1783999073758478650,beautific,0.00,355374962,de91420a-9375-4477-b44f-c072e1604f2c
3468885,2019-12-30 12:36:57+00:00,cart,5916983,1783999073758478650,beautific,27.52,592007896,84475418-083b-4a8e-ab7f-8d38994b1f12
3470599,2019-12-30 12:59:20+00:00,view,5916983,1783999073758478650,beautific,27.52,477270774,e5670c53-8674-4725-86e2-dc6bfd39c04c
3498570,2019-12-30 21:43:54+00:00,view,5916983,1783999073758478650,beautific,27.52,583884978,d6f873d5-7935-454c-9849-9fbf1b05fd53


In [ ]:
cosmetic_df.isna().sum()

,0
event_time,0
event_type,0
product_id,0
category_id,0
brand,1503141
price,0
user_id,0
user_session,779


- brand, user_session 컬럼의 결측치는 N/A 값으로 대체하는 방식으로 처리하겠다.

In [ ]:
cosmetic_df[['brand', 'user_session']] = cosmetic_df[['brand', 'user_session']].fillna('N/A')

In [ ]:
cosmetic_df.isna().sum()

,0
event_time,0
event_type,0
product_id,0
category_id,0
brand,0
price,0
user_id,0
user_session,0


## 이상치 확인

In [ ]:
cosmetic_df.describe().apply(lambda x: x.apply('{0:.2f}'.format))

,product_id,category_id,price,user_id
count,3533286.00,3533286.00,3533286.00,3533286.00
mean,5473053.80,1555023012945680128.00,8.87,522331835.07
std,1331331.39,168926187070693152.00,19.86,84948189.60
min,3752.00,1487580004807082752.00,-79.37,1180452.00
25%,5726191.00,1487580006300255232.00,2.06,486682994.00
50%,5811429.00,1487580008447738880.00,4.21,556649586.00
75%,5859462.00,1487580013581566208.00,7.14,582801855.00
max,5917178.00,2235524499636224256.00,327.78,595414541.00


- 수치형 데이터 중 product_id, category_id, user_id는 id 성격의 데이터이기 때문에 기술 통계량을 확인하는 것이 무의해보인다.
- price 컬럼의 경우 음수(-) 값이 존재하므로 파악이 필요하다.

In [ ]:
# price가 음수인 경우
len(cosmetic_df.query("price < 0"))

18

In [ ]:
cosmetic_df.query("price < 0")['product_id'].unique()

array([5670257, 5716859, 5716857, 5716855, 5716861])

- 위 5가지 상품의 경우 가격이 음수이므로, 주문 취소나 사은품 지급 등으로 인한 비용 처리 목적으로 음수가 들어갔을 수 있지만, 주어진 데이터만으로는 정확한 이유를 확인하기 어렵다. 따라서, 가격이 음수인 데이터는 분석에서 제외하겠다.

In [ ]:
# price가 0인 경우
len(cosmetic_df.query("price == 0"))

7589

In [ ]:
cosmetic_df.query("price == 0")['product_id'].nunique()

2156

In [ ]:
cosmetic_df.query("product_id == 5916983")

,event_time,event_type,product_id,category_id,brand,price,user_id,user_session
3374835,2019-12-29 07:11:47+00:00,view,5916983,1783999073758478650,beautific,0.00,410144142,57484fc7-f6a4-45ba-b95a-9116ead6d96f
3397074,2019-12-29 13:00:50+00:00,cart,5916983,1783999073758478650,beautific,0.00,355374962,de91420a-9375-4477-b44f-c072e1604f2c
3397086,2019-12-29 13:00:55+00:00,remove_from_cart,5916983,1783999073758478650,beautific,0.00,355374962,de91420a-9375-4477-b44f-c072e1604f2c
3397087,2019-12-29 13:00:55+00:00,remove_from_cart,5916983,1783999073758478650,beautific,0.00,355374962,de91420a-9375-4477-b44f-c072e1604f2c
3397111,2019-12-29 13:01:02+00:00,view,5916983,1783999073758478650,beautific,0.00,355374962,de91420a-9375-4477-b44f-c072e1604f2c
3468885,2019-12-30 12:36:57+00:00,cart,5916983,1783999073758478650,beautific,27.52,592007896,84475418-083b-4a8e-ab7f-8d38994b1f12
3470599,2019-12-30 12:59:20+00:00,view,5916983,1783999073758478650,beautific,27.52,477270774,e5670c53-8674-4725-86e2-dc6bfd39c04c
3498570,2019-12-30 21:43:54+00:00,view,5916983,1783999073758478650,beautific,27.52,583884978,d6f873d5-7935-454c-9849-9fbf1b05fd53


In [ ]:
cosmetic_df['price'] = cosmetic_df.groupby('product_id')['price'].transform(lambda x: x.replace(0, x[x != 0].mean()))

In [ ]:
cosmetic_df.query("product_id == 5916983")

,event_time,event_type,product_id,category_id,brand,price,user_id,user_session
3374835,2019-12-29 07:11:47+00:00,view,5916983,1783999073758478650,beautific,27.52,410144142,57484fc7-f6a4-45ba-b95a-9116ead6d96f
3397074,2019-12-29 13:00:50+00:00,cart,5916983,1783999073758478650,beautific,27.52,355374962,de91420a-9375-4477-b44f-c072e1604f2c
3397086,2019-12-29 13:00:55+00:00,remove_from_cart,5916983,1783999073758478650,beautific,27.52,355374962,de91420a-9375-4477-b44f-c072e1604f2c
3397087,2019-12-29 13:00:55+00:00,remove_from_cart,5916983,1783999073758478650,beautific,27.52,355374962,de91420a-9375-4477-b44f-c072e1604f2c
3397111,2019-12-29 13:01:02+00:00,view,5916983,1783999073758478650,beautific,27.52,355374962,de91420a-9375-4477-b44f-c072e1604f2c
3468885,2019-12-30 12:36:57+00:00,cart,5916983,1783999073758478650,beautific,27.52,592007896,84475418-083b-4a8e-ab7f-8d38994b1f12
3470599,2019-12-30 12:59:20+00:00,view,5916983,1783999073758478650,beautific,27.52,477270774,e5670c53-8674-4725-86e2-dc6bfd39c04c
3498570,2019-12-30 21:43:54+00:00,view,5916983,1783999073758478650,beautific,27.52,583884978,d6f873d5-7935-454c-9849-9fbf1b05fd53


In [ ]:
# price가 0인 경우
len(cosmetic_df.query("price == 0"))

0

In [ ]:
# price가 양수인 데이터만 필터링
cosmetic_df = cosmetic_df.query("price > 0")

## 중복값 확인

In [ ]:
cosmetic_df.shape

(3531840, 8)

In [ ]:
# 중복값 개수 확인
cosmetic_df.duplicated().sum()

np.int64(183858)

In [ ]:
183858 / 3531840 * 100

5.205728458820332

In [ ]:
# 중복값 제거
cosmetic_df = cosmetic_df.drop_duplicates()

In [ ]:
# 중복값 개수 확인
cosmetic_df.duplicated().sum()

np.int64(0)

In [ ]:
cosmetic_df.shape

(3347982, 8)

# 퍼널 분석용 데이터 준비

- 퍼널 분석(Funnel Analysis) : 고객이 특정 목표에 도달하기까지 거치는 일련의 단계를 분석하는 기법
- 퍼널 분석을 활용하면 특정 목표에 도달하기까지 어떤 단계에 개선이 필요한지 확인할 수 있어 웹 사이트나 앱 내의 화면 기획 시 주요 참고 자료로 활용 가능

In [ ]:
cosmetic_df.head()

,event_time,event_type,product_id,category_id,brand,price,user_id,user_session
0,2019-12-01 00:00:00+00:00,remove_from_cart,5712790,1487580005268456287,f.o.x,6.27,576802932,51d85cb0-897f-48d2-918b-ad63965c12dc
1,2019-12-01 00:00:00+00:00,view,5764655,1487580005411062629,cnd,29.05,412120092,8adff31e-2051-4894-9758-224bfa8aec18
2,2019-12-01 00:00:02+00:00,cart,4958,1487580009471148064,runail,1.19,494077766,c99a50e8-2fac-4c4d-89ec-41c05f114554
3,2019-12-01 00:00:05+00:00,view,5848413,1487580007675986893,freedecor,0.79,348405118,722ffea5-73c0-4924-8e8f-371ff8031af4
4,2019-12-01 00:00:07+00:00,view,5824148,1487580005511725929,N/A,5.56,576005683,28172809-7e4a-45ce-bab0-5efa90117cd5


In [ ]:
# user_id, product_id, event_time을 기준으로 데이터 정렬
cosmetic_df = cosmetic_df.sort_values(by=['user_id', 'product_id', 'event_time'])

In [ ]:
# user_id, product_id 기준으로 첫 이벤트
temp = cosmetic_df.groupby(['user_id', 'product_id']).first()

In [ ]:
temp.head()

event_time event_type          category_id  \
user_id product_id                                                             
1180452 5881337    2019-12-28 14:32:56+00:00       view  1487580012096782476   
2963072 5656122    2019-12-22 13:04:49+00:00       view  1487580009026551821   
        5677604    2019-12-22 13:11:02+00:00       cart  1487580010100293687   
        5700025    2019-12-22 13:09:11+00:00       view  1487580010100293687   
        5707747    2019-12-22 12:50:45+00:00       view  1487580009051717646   

                     brand  price                          user_session  
user_id product_id                                                       
1180452 5881337      fedua  25.40  a4818e6d-9069-4aa8-8731-572ac266283f  
2963072 5656122        N/A  23.57  3bf2bbbb-1a32-4f06-a263-a3f49db74750  
        5677604        N/A   1.27  3bf2bbbb-1a32-4f06-a263-a3f49db74750  
        5700025     runail   1.90  3bf2bbbb-1a32-4f06-a263-a3f49db74750  
        5707747        N/A  73.02  3bf2bbbb-1a32-4f06-a263-a3f49db74750

In [ ]:
# user_id - product_id 조합의 첫 이벤트가 view인 경우만
target_index = temp.query("event_type == 'view'").index

In [ ]:
target_index

MultiIndex([(  1180452, 5881337),
            (  2963072, 5656122),
            (  2963072, 5700025),
            (  2963072, 5707747),
            (  2963072, 5746011),
            (  2963072, 5774240),
            (  2963072, 5841190),
            (  2963072, 5875320),
            (  2963072, 5896414),
            (  2963072, 5898229),
            ...
            (595413244, 5863531),
            (595413309, 5893983),
            (595413447, 5861554),
            (595413503, 5831578),
            (595413665, 5875992),
            (595413843, 5696330),
            (595413976, 5745711),
            (595414210, 5886338),
            (595414257, 5812934),
            (595414541,   59975)],
           names=['user_id', 'product_id'], length=1167266)

In [ ]:
# user_id - product-id 조합의 첫 이벤트가 view인 경우
cosmetic_df = cosmetic_df.set_index(['user_id', 'product_id']).loc[target_index].reset_index()

# 퍼널 분석

- 목표 : 고객의 구매(purchase) 촉진
- 퍼널 단계 : view -> cart -> purchase

## 단계별 전환율

In [ ]:
cosmetic_df.head()

,user_id,product_id,event_time,event_type,category_id,brand,price,user_session
0,1180452,5881337,2019-12-28 14:32:56+00:00,view,1487580012096782476,fedua,25.40,a4818e6d-9069-4aa8-8731-572ac266283f
1,2963072,5656122,2019-12-22 13:04:49+00:00,view,1487580009026551821,N/A,23.57,3bf2bbbb-1a32-4f06-a263-a3f49db74750
2,2963072,5700025,2019-12-22 13:09:11+00:00,view,1487580010100293687,runail,1.90,3bf2bbbb-1a32-4f06-a263-a3f49db74750
3,2963072,5707747,2019-12-22 12:50:45+00:00,view,1487580009051717646,N/A,73.02,3bf2bbbb-1a32-4f06-a263-a3f49db74750
4,2963072,5707747,2019-12-22 12:53:12+00:00,view,1487580009051717646,N/A,73.02,3bf2bbbb-1a32-4f06-a263-a3f49db74750


In [ ]:
# 각 단계별 unique한 product_id - user_id 조합 집계
user_counts = cosmetic_df.groupby(['event_type', 'product_id'])['user_id'].nunique()

In [ ]:
user_counts

event_type  product_id
cart        3762          67
            3763           2
            3774          19
            3928           4
            3936           7
                          ..
view        5917159        2
            5917162        1
            5917163        2
            5917175        4
            5917177        5
Name: user_id, Length: 101246, dtype: int64

In [ ]:
# 각 단계별 유저 수 집계
user_counts = user_counts.groupby('event_type').sum().reset_index()

In [ ]:
# 컬럼명 변경(user_id -> users)
user_counts = user_counts.rename(columns={'user_id' : 'users'})

In [ ]:
# 'remove_from_cart' 단계 제외
user_counts = user_counts.query("event_type != 'remove_from_cart'")

In [ ]:
# 퍼널 단계 순으로 정렬(유저 수가 많은 단계에서 적은 단계 순으로 정렬)
user_counts = user_counts.sort_values(by='users', ascending=False)

In [ ]:
user_counts

,event_type,users
3,view,1167266
0,cart,145919
1,purchase,53250


In [ ]:
# 퍼널 시각화
px.funnel(user_counts,
          x='users',
          y='event_type')

In [ ]:
user_counts

,event_type,users
3,view,1167266
0,cart,145919
1,purchase,53250


In [ ]:
# 전환율 계산
user_counts['conversion_rate(%)'] = user_counts['users'] / user_counts['users'].shift(1) * 100

In [ ]:
# 이탈률 계산
user_counts['churn_rate(%)'] = 100 - user_counts['conversion_rate(%)']

In [ ]:
user_counts

,event_type,users,conversion_rate(%),churn_rate(%)
3,view,1167266,NaN,NaN
0,cart,145919,12.500921,87.499079
1,purchase,53250,36.492849,63.507151


- 유저 관점에서 상품 탐색의 흐름을 봤을 때, view에서 cart로 넘어가는 단계에서 약 87.5%가 이탈하고, cart에서 purchase로 넘어가는 단계에서는 약 63.5%가 이탈하고 있다.

## view 100% 기준 전환율

In [ ]:
user_counts

,event_type,users,conversion_rate(%),churn_rate(%)
3,view,1167266,NaN,NaN
0,cart,145919,12.500921,87.499079
1,purchase,53250,36.492849,63.507151


In [ ]:
# view 100%
user_counts.iloc[0, 1]

np.int64(1167266)

In [ ]:
# view를 100%로 두고 전환율 계산
user_counts['conversion_rate_total'] = user_counts['users'] / user_counts.iloc[0, 1]

In [ ]:
user_counts

,event_type,users,conversion_rate(%),churn_rate(%),conversion_rate_total
3,view,1167266,NaN,NaN,1.000000
0,cart,145919,12.500921,87.499079,0.125009
1,purchase,53250,36.492849,63.507151,0.045619


In [ ]:
# 퍼널 시각화
fig = px.funnel(user_counts,
                x='conversion_rate_total',
                y='event_type')
fig.update_traces(texttemplate="%{value:.2%}")

- 상품을 본 후, 장바구니에 담기까지의 전환율은 약 12.5%이고,
- 상품을 본 후, 구매까지 이어지는 전환율은 약 4.6%임을 확인할 수 있다.

## 가격대별 분석

- 상품의 가격대에 따라 전환율에 차이가 있는지 확인하고자 한다.

In [ ]:
cosmetic_df.head()

,user_id,product_id,event_time,event_type,category_id,brand,price,user_session
0,1180452,5881337,2019-12-28 14:32:56+00:00,view,1487580012096782476,fedua,25.40,a4818e6d-9069-4aa8-8731-572ac266283f
1,2963072,5656122,2019-12-22 13:04:49+00:00,view,1487580009026551821,N/A,23.57,3bf2bbbb-1a32-4f06-a263-a3f49db74750
2,2963072,5700025,2019-12-22 13:09:11+00:00,view,1487580010100293687,runail,1.90,3bf2bbbb-1a32-4f06-a263-a3f49db74750
3,2963072,5707747,2019-12-22 12:50:45+00:00,view,1487580009051717646,N/A,73.02,3bf2bbbb-1a32-4f06-a263-a3f49db74750
4,2963072,5707747,2019-12-22 12:53:12+00:00,view,1487580009051717646,N/A,73.02,3bf2bbbb-1a32-4f06-a263-a3f49db74750


In [ ]:
# 상품별 가격 데이터 정리
temp = cosmetic_df[['product_id', 'price']].drop_duplicates()

In [ ]:
temp['price'].quantile([0.25, 0.5, 0.75])

,price
0.25,2.78
0.50,4.86
0.75,8.41


In [ ]:
# 가격대 함수
def group_price(x):
    if x <= 3:
        return '~3'
    elif x <= 5:
        return '3~5'
    elif x <= 8:
        return '5~8'
    else:
        return '8~'

In [ ]:
cosmetic_df['price_range'] = cosmetic_df['price'].apply(group_price)

In [ ]:
cosmetic_df.head()

,user_id,product_id,event_time,event_type,category_id,brand,price,user_session,price_range
0,1180452,5881337,2019-12-28 14:32:56+00:00,view,1487580012096782476,fedua,25.40,a4818e6d-9069-4aa8-8731-572ac266283f,8~
1,2963072,5656122,2019-12-22 13:04:49+00:00,view,1487580009026551821,N/A,23.57,3bf2bbbb-1a32-4f06-a263-a3f49db74750,8~
2,2963072,5700025,2019-12-22 13:09:11+00:00,view,1487580010100293687,runail,1.90,3bf2bbbb-1a32-4f06-a263-a3f49db74750,~3
3,2963072,5707747,2019-12-22 12:50:45+00:00,view,1487580009051717646,N/A,73.02,3bf2bbbb-1a32-4f06-a263-a3f49db74750,8~
4,2963072,5707747,2019-12-22 12:53:12+00:00,view,1487580009051717646,N/A,73.02,3bf2bbbb-1a32-4f06-a263-a3f49db74750,8~


In [ ]:
# 각 단계별 unique한 product_id - user_id 조합 집계
price_funnel = cosmetic_df.groupby(['price_range', 'event_type', 'product_id'])['user_id'].nunique()

In [ ]:
price_funnel = price_funnel.groupby(['price_range', 'event_type']).sum().reset_index()

In [ ]:
price_funnel

,price_range,event_type,user_id
0,3~5,cart,33321
1,3~5,purchase,11685
2,3~5,remove_from_cart,17717
3,3~5,view,237058
4,5~8,cart,31101
5,5~8,purchase,10586
6,5~8,remove_from_cart,15357
7,5~8,view,238430
8,8~,cart,31726
9,8~,purchase,11138


In [ ]:
# 컬럼명 변경(user_id -> users)
price_funnel = price_funnel.rename(columns={'user_id':'users'})

In [ ]:
# 'remove_from_cart' 단계 제외
price_funnel = price_funnel.query("event_type != 'remove_from_cart'")

In [ ]:
# 정렬
price_funnel = price_funnel.groupby('price_range').apply(lambda x: x.sort_values(by='users', ascending=False)).reset_index(drop=True)

/tmp/ipykernel_1594/1959501948.py:2: DeprecationWarning:

DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.



In [ ]:
price_funnel

,price_range,event_type,users
0,3~5,view,237058
1,3~5,cart,33321
2,3~5,purchase,11685
3,5~8,view,238430
4,5~8,cart,31101
5,5~8,purchase,10586
6,8~,view,375656
7,8~,cart,31726
8,8~,purchase,11138
9,~3,view,316323


In [ ]:
# view를 100%로 두고 전환율 계산
price_funnel['conversion_rate'] = price_funnel['users'] / price_funnel.groupby('price_range')['users'].transform('first')

In [ ]:
price_funnel

,price_range,event_type,users,conversion_rate
0,3~5,view,237058,1.000000
1,3~5,cart,33321,0.140561
2,3~5,purchase,11685,0.049292
3,5~8,view,238430,1.000000
4,5~8,cart,31101,0.130441
5,5~8,purchase,10586,0.044399
6,8~,view,375656,1.000000
7,8~,cart,31726,0.084455
8,8~,purchase,11138,0.029649
9,~3,view,316323,1.000000


In [ ]:
# 퍼널 시각화
fig = px.funnel(price_funnel,
                x='conversion_rate',
                y='event_type',
                facet_col='price_range',
                color='price_range',
                category_orders={'price_range':['~3', '3~5', '5~8', '8~']})
fig.update_traces(texttemplate="%{value:.2%}")

- 가격대가 높은 상품군에서 구매로까지 이어지는 전환율이 상대적으로 낮게 나타나는 것을 확인할 수 있다.

In [ ]:
# 가격대별 이벤트별 유저 수
pivot_temp = price_funnel.pivot(index='price_range', columns='event_type', values='users')

In [ ]:
pivot_temp

event_type,cart,purchase,view
price_range,,,
3~5,33321,11685,237058
5~8,31101,10586,238430
8~,31726,11138,375656
~3,49778,19841,316323


In [ ]:
# view -> cart 전환율, cart -> purchase 전환율 계산
pivot_temp['view_to_cart'] = pivot_temp['cart'] / pivot_temp['view'] * 100
pivot_temp['cart_to_purchase'] = pivot_temp['purchase'] / pivot_temp['cart'] * 100

In [ ]:
pivot_temp

event_type,cart,purchase,view,view_to_cart,cart_to_purchase
price_range,,,,,
3~5,33321,11685,237058,14.056054,35.067975
5~8,31101,10586,238430,13.044080,34.037491
8~,31726,11138,375656,8.445493,35.106852
~3,49778,19841,316323,15.736447,39.858974


In [ ]:
# index 순서 지정
pivot_temp.reindex(index=['~3', '3~5', '5~8', '8~'])

event_type,cart,purchase,view,view_to_cart,cart_to_purchase
price_range,,,,,
~3,49778,19841,316323,15.736447,39.858974
3~5,33321,11685,237058,14.056054,35.067975
5~8,31101,10586,238430,13.044080,34.037491
8~,31726,11138,375656,8.445493,35.106852


- 가격대가 높아질수록 view -> cart로의 전환율이 떨어지고, cart -> purchase로의 전환율은 3 이하 그룹과 3 초과 그룹 간의 차이는 존재하지만, 3 초과 가격대에서는 cart -> purchase로의 전환율에 큰 차이가 나지 않음을 확인할 수 있다.



# 결론

**분석 결과 요약**
- 상품을 확인(view)한 후 장바구니(cart)에 담는 전환율은 약 12.5%, 상품을 확인(view)한 후 상품이 구매(purchase)까지 이어지는 전환율은 약 4.6%임을 확인할 수 있다.
- 가격대가 높아질수록 상품이 구매(purchase)까지 이어지는 전환율이 감소하는 것을 확인할 수 있다.
- 가격대가 높아질수록 view -> cart로의 전환율이 떨어지고, cart -> purchase로의 전환율은 3 이하 그룹과 3 초과 그룹 간의 차이는 존재하나 3초과 가격대에서는 cart -> purchase로의 전환율에 별다른 차이가 나타나지 않음을 확인할 수 있다.

**개선 전략 제안**
- 가격대가 높은 상품일수록 할인 혜택 등을 강조하여 view -> cart로의 전환율을 1차적으로 높일 수 있도록 신경을 써야 한다.
- 가격대가 3 이하인 상품이라면, cart -> purchase로의 전환율이 다른 가격대와 비교하였을 때 상대적으로 높으므로 카트 내 상품에 대한 알림을 통해 구매를 독려하는 전략을 시도해볼 수 있다.